### ISSUE TITLE
Who stops the process? GC or libscanmem?

### ISSUE THREAD

**Issue by korcankaraokcu** [scanmem/scanmem#198]
Who stops the process when doing the scan? GameConqueror or libscanmem? This is a very important issue because if libscanmem automatically stops the process when a scan starts then continues it, it won't go well with debuggers. Think a situation where user is debugging something and in the middle of debug, he wants to scan the memory. When the scan is finished the process will continue to run automatically and unfortunately, it'll ruin the entire debugging session.

I also think that GC or scanmem should have a option like "Don't continue after scanning", allowing the user to handle signal passing

**Comment by sriemer** [comment]
libscanmem does ptrace(PTRACE_ATTACH,...) in ptrace.c. GC just sends scanmem commands to the libscanmem backend. It detaches again once the scan is over.

**Comment by sriemer** [comment]
GC has that auto-value updater calling the `update` command all the time. People want this feature and it can't be disabled for now. Use scanmem on the command line if you need a debugger in parallel.

**Comment by 12345ieee** [comment]
On this topic, I read about `PTRACE_SEIZE`, that should be able to decouple attaching from sending `SIGSTOP`.

I wanted freezing (aka `SIGSTOP`) the process to be independent from scanning the process, but I don't know if this is feasible.

An expert's opinion, @sriemer ?

**Comment by sriemer** [comment]
Sure, freezing the process can be done without scanning but libscanmem has to be changed for this. Maybe new commands `freeze` and `unfreeze` could help but then scanmem has to remember that the target process is already frozen for not doing that again.

Question is: What should happen after the scan then?
Unfreeze or stay attached? I would say "stay attached".

**Comment by 12345ieee** [comment]
I would split the current `attach&freeze` and `detach&unfreeze` (both unrelated to `scan`) in: 
* `attach-only`
* `freeze/unfreeze` callable at anytime
* `detach-only`

Clearly you need an `isfrozen` bool.

Then you make a scanmem switch to retain the old behaviour (default to it maybe), which transforms `scan` in `freeze-scan-unfreeze`.

I understand this is a rather big change, but it's something I'd really like for v0.17 and something that would ease downstream devs' flexibility.

**Comment by korcankaraokcu** [comment]
@12345ieee's idea is the most generalized form of it, I also approve it. But for libscanmem, how about also making attach&detach callable at any time? Because a debugger would only need scanning, it would handle freeze/unfreeze and attach/detach by itself.

**Comment by 12345ieee** [comment]
[I'm doing a clean-up of the dangling issues] So, this question got definitely sidetracked.

I'll look into `PTRACE_SEIZE`, to be able to implement my idea, but this is for another issue.

@korcankaraokcu , currently attach/detach **are** callable by themselves, via `sm_attach/detach()`, for example the scanning functions call these automatically at the start.
These functions are just thin wrappers over `ptrace(correct_flags...)`.

If I understand you, you have already attached to the target from another process, and don't want `libscanmem` to restart your process.

This looks not possible to me, `ptrace_detach` seems to require the process to be stopped (ok), but then restarts it with no recourse. Of course you can deliver a SIGSTOP as soon as it's detached.

Now, IF you have `/proc` support, you don't really need all the fancy `ptrace()` work, you just need to (optionally) send a SIGSTOP and everything works.

So, to answer:
* it's `libscanmem` that attaches/detaches
* while the state of the target during the scan can be changed, detach always proceeds from stopped to running (at least, that's what the somewhat obscure docs seem to say)
* a partial rewrite of some scanmem internals could get rid of the `ptrace()` calls if you have `/proc` support, replacing them with `kill(SIGCONT/STOP)`, completely eliminating the above problem (and several other problems in all the modern distros)

@korcankaraokcu , comments?

**Comment by korcankaraokcu** [comment]
Thanks for the detailed answer. I guess I can rewrite the required internals. Integrating scanmem with gdb is one of the last steps in roadmap anyway, so there'll be a plenty of time. I'll check scanmem's situation again when the time comes. Let's see what will future bring.
